### ISSUE TITLE
Foreign types can be passed to generics, even though the documentation states this is invalid

### ISSUE THREAD

**Issue by yorickpeterse** [inko-lang/inko#731]
Since we don't generate unique shapes for all foreign types (e.g. we don't have shapes for `Int32` and `Int16`), we're supposed to disallow passing foreign types to generic type parameters (e.g. `Array[Int32]` shouldn't be possible). However, the compiler fails to correctly enforce this, resulting in code such as `[42 as Int32]` being valid.

We should fix this by just generic the correct shapes for foreign types. To do so, `Shape::Int` and `Shape::Float` should be changed to wrap an `u8` that specifies the size in bits. We then have to update `Shape::identifier` to emit the following names:

- `Int8`: `i8`
- `Int16`: `i16`
- `Int32`: `i32`
- `Int64` / `Int`: `i64`
- `UInt8`: `u8`
- `UInt16`: `u16`
- `UInt32`: `u32`
- `UInt64`: `u64`
- `Float32`: `f32`
- `Float64` / `Float`: `f64`

In addition, `TypeSpecializer` needs to be updated, and the LLVM IR should be typed correctly for these types. 

For `Array` we also need some sort of compile-time expression that gets the type size of a generic type parameter, such that we can use this to determine the amount of memory to allocate, instead of assuming this is always 8 bytes per value.

**Comment by yorickpeterse** [comment]
For this to work, we first need an expression that takes a type and returns its size or alignment. This is needed because `Array` currently assumes values are always 8 bytes, but that won't be the case anymore if we allow C types to be stored in arrays (and other generic types).

**Comment by yorickpeterse** [comment]
For `size_of()` and equivalent, I've been toying with different syntaxes but they all suck. One option I've been thinking about is making types valid expressions, with their return values being a value containing the size and alignment. So `Array[Int]` would return essentially `(size, alignment)`. The downside is that we don't always need everything included, and if the data is heap allocated it's too expensive.

**Comment by yorickpeterse** [comment]
Another option is to do this: we _do_ allow types in expressions, but similar to class names we only allow them to be used as receivers. We then special-case the calls `size` and `alignment` when their receivers are types. This however is also a bit hacky, so I'm not a big fan of it.

**Comment by yorickpeterse** [comment]
The syntax `Type.method` is ambiguous: given for example `String.size`, it's not clear if `size` is a static method, or a compile-time method.

One option is to introduce the syntax `type(X)` which returns some type object, such that you can then do `type(X).size` to get the size.

**Comment by yorickpeterse** [comment]
`type()` would just return the class of the type, which works out for generic types since they're specialized to have their own classes. Classes also already store the size in bytes of their instances, and we can extend them to also include the alignment.

Worth mentioning: it seems we no longer use the instance size still stored in classes, so if we don't implement this we should get rid of that.

**Comment by yorickpeterse** [comment]
What I'm currently leaning towards is this:

We add support for the syntax `type(X)` (e.g. `type(Array[Int])`). This returns a `Type`, defined as follows:

```inko
class extern Type {
  let @size: Int
}
```

For now we'll need to use `class extern` to ensure there's no space wasted for an object header, and to avoid heap allocations. Additional information such as the alignment won't be included for now.

In `std.array` we then replace references to `VALUE_SIZE` with `type(T).size`. 

In the future, when we allow user-defined value/inline types we'd replace the `class extern` with `class inline`. The difference between the two is that `class inline` can be cast to a trait by boxing it, and the compiler might re-arrange fields, while for `class extern` neither is the case.

**Comment by yorickpeterse** [comment]
To add to the above: for the `type()` expression, the returned value should take into account whether the type is heap or stack allocated. That is, the size should be the size of a reference to the type on the stack, which for most types would be 8 bytes.

**Comment by yorickpeterse** [comment]
Perhaps an easier way is to add a `_INKO.array_type_parameter_size(array)` intrinsic of sorts, returning the size of the array's type parameter. While runtime type information to some degree is useful, I think it's too early to add this now. I also don't want to expose too much information about object layouts and sizes, as that makes it more difficult to change it over time (e.g. as users may come to depend on specific sizes and what not).

**Comment by yorickpeterse** [comment]
I played around with a few different approaches. Based on this I decided it's best to _not_ allow C types in generics as originally intended. As such types aren't proper Inko types, you can't call any methods on them. This means that in practice, you wouldn't actually use C types in generics other than to move them in and out of collections, but for that a `x as Int` is sufficient (and better).

Allowing C types in generics also essentially "clobbers" the language and compiler, as we have to support all the different integer/float sizes and their signed status, even though the language's official integer type is just `i64`.

In other words, we'd be making changes for something that wouldn't be useful.

So we'll do the following:

1. Fix type-checking to not allow C types in generics
2. Fix `Array` to not use a hard-coded size for each value, in preparation for future support of stack allocated custom types

**Comment by yorickpeterse** [comment]
Note to self: we should disallow regular Inko types inside `class extern` types, as this can be used to violate the single-ownership constraints, since `class extern` types are value types.

**Comment by yorickpeterse** [comment]
This is fixed by 119e77bc25edc6962c5267a03395a7f54aa1ec94.
